src/main/java/com/google/sps/data/Journey.java

```diff
@@ -79,45 +87,181 @@ public int getNumberOfWaypoints() {
+    int allWaypointsConfig = (1 << numberOfWaypoints) - 1;
+    int minTime = bestTime.get((1 << numberOfWaypoints)).get(0).first;
+    for (int waypointIndex = 0; waypointIndex < numberOfWaypoints; ++waypointIndex) {
+      if (bestTime.get(allWaypointsConfig).get(waypointIndex).first < minTime) {
+        minTime = bestTime.get(allWaypointsConfig).get(waypointIndex).first;
+      }
+    }
+
+    long deliverySlotDurationInSeconds = Duration.between(deliverySlot.getStartTime().toInstant(),
+        deliverySlot.getEndTime().toInstant()).toMillis() / 1000;
+
+    return (minTime <= (int)deliverySlotDurationInSeconds);
   }
 
   /**
    * Computes and returns the optimal order in which points should be visited starting from the
    * last point to the fist one (the start Point).
    */
-  public ArrayList<Point> findOptimalOrderForWaypoints() {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public ArrayList<Point> findOptimalOrderForWaypoints() throws ApiException, DataNotFoundException, IOException, InterruptedException {
+    ArrayList<ArrayList<Pair>> bestTime = findOptimalTimePaths();
+    ArrayList<Point> orderedWaypoints = new ArrayList<>();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    // the last state of the matrix has all waypoints set to 1
+    int currentState =  (1 << numberOfWaypoints) - 1;
+    
+    int minTime = bestTime.get((1 << numberOfWaypoints)).get(0).first;
+    // currentWaypoint is the one for which bestTime[currentState][currentWaypoint] is minimal
+    int currentWaypoint = 0;
+    for (int waypointIndex = 1; waypointIndex < numberOfWaypoints; ++waypointIndex) {
+      if (bestTime.get(currentState).get(waypointIndex).first < minTime) {
+        minTime = bestTime.get(currentState).get(waypointIndex).first;
+        currentWaypoint = waypointIndex;
+      }
+    }
+    while (currentState != 0) {
+      orderedWaypoints.add(waypoints.get(currentWaypoint));
+      int prevState = currentState ^ (1 << currentWaypoint);
+      int prevWaypoint = bestTime.get(currentState).get(currentWaypoint).second;
+      currentWaypoint = prevWaypoint;
+      currentState = prevState;
+    }
+    // add the start of the journey at the end of orderedWaypoints
+    orderedWaypoints.add(start);
+    return orderedWaypoints;
+  }
+
+  /**
+   * Creates and initializes bestTime matrix, a matrix where bestTime[config][lastWaypoint] = the
+   * minimum time to visit all points with bits set to 1 in config such that the last visited point
+   * is lastWaypoint; config is a bitmask where bits set to 1 represent currently visited waypoints
+   */
+  private ArrayList<ArrayList<Pair>> initializeBestTimeMatrix() throws ApiException, IOException, InterruptedException, DataNotFoundException {
+    ArrayList<ArrayList<Pair>> bestTime = new ArrayList<>();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    for (int config = 0; config <= (1 << numberOfWaypoints); ++config) {
```

Comment:
I'm sorry that was a bug I missed: I changed it to bestTime[(1 << numberOfWaypoints) - 1] . Test passed because there was no test for which the first added point should be visited last. I will add one test for that.

For numberOfWaypoints = 2, allWaypointsConfig = 3 = 11(2) meaning that waypoints with indexes 0 and 1 must be visited. The indexes of waypoints are between 0 and numberOfWaypoint - 1.